### ISSUE TITLE
skaffold dev: deploy phase hangs because of a dependency that is further down the queue

### ISSUE THREAD

**Issue by gregwhorley** [GoogleContainerTools/skaffold#9126]
### Expected behavior
TL;DR -- Running Skaffold 2.0 in dev mode will deploy all resources without hanging, just like it does with our Skaffold 1.0 configuration. The kustomize deploy step in 1.0 always seemed to apply the needed resources before the helm deploy steps were executed and `skaffold dev` worked for our use case. In 2.0, the kubectl deploy step that replaced kustomize does not run first, instead helm releases are deployed first and cause the deploy step to hang.

I am trying to upgrade a project's skaffold config from `v2beta29` (skaffold 1.0) to latest `v4` (skaffold 2.0).  I made all of the necessary syntax changes that conform to the new schema and I can perform build and render operations successfully. Here's my problem...

Our devs rely on `skaffold dev` for local testing and debugging, and we were able to construct a skaffold 1.0 configuration that allows each developer to have their own unique set of k8s resources in the same namespace through an exhaustive set of patches that basically rename everything in our kustomize and skaffold configurations by adding a prefix. We use a combination of kustomize and helm in our deployment pipeline, where the former manages several configmaps/secrets that the latter depends on. Here's where the upgrade to skaffold 2.0 becomes an issue in our case.

If I have an `application.skaffold.yaml` that builds and deploys our first party app:
```yaml
apiVersion: skaffold/v4beta7
kind: Config
metadata:
  name: application
build:
  local:
    push: false
  artifacts:
    - image: application
deploy:
  helm:
    releases:
      - name: "application"
        remoteChart: oci://our-first-party-app-chart
        valuesFiles:
          - "redacted"
        setValueTemplates:
          image.repository: "{{.IMAGE_FULLY_QUALIFIED_application}}"
          fullnameOverride: "application"
          secretName: "dbsecrets" # <--- This is managed by kustomize
          configMapName: "config" # <--- This is managed by kustomize
profiles:
  - name: dev
    manifests:
      kustomize:
        paths:
          - ../redacted
    deploy:
      kubectl: {}
    patches:
      - op: replace
        path: /build/artifacts/0/image
        value: "application-dev"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/image.repository
        value: "{{.IMAGE_FULLY_QUALIFIED_application_dev}}"
      - op: replace
        path: /deploy/helm/releases/0/name
        value: "{{.PREFIX}}-application"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/fullnameOverride
        value: "{{.PREFIX}}-application"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/secretName      # <--- also renamed in kustomize
        value: "{{.PREFIX}}-dbsecrets"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/configMapName   # <--- also renamed in kustomize
        value: "{{.PREFIX}}-config"
```

The application mounts secrets and configmaps that are defined in `kustomization.yaml`:
```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
secretGenerator:
  - name: dbsecrets
    literals:
      - SECRET1=<SECRET_1>
configmapGenerator:
  - name: config
    literals:
      - CONFIG1=hello
patches:
  - target:
      kind: Secret
      name: dbsecrets
    patch: |-
      - op: replace
        path: /metadata/name
        value: prefix-dbsecrets
  - target:
      kind: ConfigMap
      name: config
    patch: |-
      - op: replace
        path: /metadata/name
        value: prefix-config
```

If I run `skaffold dev --profile dev -f application.skaffold.yaml -n development` the deploy phase will start with the Helm release and then hang because the dbsecrets and config resources have not yet been deployed by kubectl.

I even tried to coerce skaffold into performing the kubectl deploy before anything else by creating a new skaffold config file `dev-env.skaffold.yaml` and running `skaffold dev --profile dev -f dev-env.skaffold.yaml -n development`:

```yaml
apiVersion: skaffold/v4beta7
kind: Config
metadata:
  name: dev-env
manifests:
  kustomize:
    paths:
      - k8s/build/deploy/dev
deploy:
  kubectl:
    defaultNamespace: development
requires:
  - path: ./application/application.skaffold.yaml
profiles:
  - name: dev
```

The helm release defined in ./application/application.skaffold.yaml will consistently be first in the deployment queue.

### Actual behavior

Running Skaffold 2.0 in dev mode with our upgraded configuration will hang during deploy because deployments rolled out by Helm depend on resources created by kubectl that are further down the queue. 

### Information

- Skaffold version: 2.8.0
- Operating system: MacOS 13.5.2
- Installed via: `asdf install skaffold 2.8.0`
- Contents of skaffold.yaml:

```yaml
apiVersion: skaffold/v4beta7
kind: Config
metadata:
  name: application
build:
  local:
    push: false
  artifacts:
    - image: application
deploy:
  helm:
    releases:
      - name: "application"
        remoteChart: oci://our-first-party-app-chart
        valuesFiles:
          - "redacted"
        setValueTemplates:
          image.repository: "{{.IMAGE_FULLY_QUALIFIED_application}}"
          fullnameOverride: "application"
          secretName: "dbsecrets" # <--- This is managed by kustomize
          configMapName: "config" # <--- This is managed by kustomize
profiles:
  - name: dev
    manifests:
      kustomize:
        paths:
          - ../redacted
    deploy:
      kubectl: {}
    patches:
      - op: replace
        path: /build/artifacts/0/image
        value: "application-dev"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/image.repository
        value: "{{.IMAGE_FULLY_QUALIFIED_application_dev}}"
      - op: replace
        path: /deploy/helm/releases/0/name
        value: "{{.PREFIX}}-application"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/fullnameOverride
        value: "{{.PREFIX}}-application"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/secretName      # <--- also renamed in kustomize
        value: "{{.PREFIX}}-dbsecrets"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/configMapName   # <--- also renamed in kustomize
        value: "{{.PREFIX}}-config"
```

### Steps to reproduce the behavior

1. Start with deployment setup that has kustomize yaml and helm releases defined and is deployed successfully with skaffold 1.0
2. Upgrade skaffold config to 2.0
3. Run `skaffold dev` on new config
4. Watch deploy hang at Helm release because it depends on a resource that is managed by another deployer further down the queue

**Comment by renzodavid9** [comment]
Hey @gregwhorley, thanks for opening this issue. From your explanation, looks like our best option is, what you mention, split your `application.skaffold.yaml` to have a kustomize renderer + kubectl deployer config as a [dependency of another config](https://skaffold.dev/docs/design/config/#configuration-dependencies) that handles the helm deployer. 

That will be something like this:
```
apiVersion: skaffold/v4beta7
kind: Config

requires:
  - path: ./kustomize-config.skaffold.yaml # <- This is another Skaffold config with the kustomize renderer + kubectl deployer

deploy:
  helm:
    releases: ...
```

From the tests you did, looks like the configs dependencies are invert? Where you able to make it work? Thanks!

**Comment by gregwhorley** [comment]
Thank you for the response @renzodavid9 ! I was able to produce a workaround where the kustomize resources are rendered and deployed via skaffold CLI commands before `skaffold dev` is executed, which is all done inside of a make target. We already set up a make target for instantiating our dev sandboxes, so this was easiest for me to implement.

```
.PHONY: skaffold-dev
skaffold-dev: ecr-login-dependency helm-repo-add-dependency kustomize-builder other-dependencies
     skaffold render --profile dev -f dev-env.skaffold.yaml -o dev-env.yaml
     skaffold apply dev-env.yaml -n development
     skaffold dev --profile dev --module application,otherModule,etc --trigger manual -n development
     skaffold delete --profile dev -f dev-env.skaffold.yaml -n development
     rm -f dev-env.yaml
```

I just tried adding the dependency config option you suggested, but I couldn't get it to fit my use case because the `requires` option is not accepted inside of a profile:

```
parsing skaffold config: error parsing skaffold configuration file: unable to parse config: yaml: unmarshal errors:
  line XXX: field requires not found in type latest.Profile
```

my test skaffold config:

```yaml
apiVersion: skaffold/v4beta7
kind: Config
metadata:
  name: application
build:
  local:
    push: false
  artifacts:
    - image: application
deploy:
  helm:
    releases:
      - name: application
        remoteChart: oci://repo/our-first-party-chart
        valuesFiles:
          - redacted
        setValueTemplates:
          image.repository: "{{.IMAGE_FULLY_QUALIFIED_application}}"
          fullnameOverride: "reporting-service"
          secretName: "dbsecrets"
          configMapName: "config"
profiles:
  - name: dev
    requires:
      - path: ../dev-env.skaffold.yaml
    patches:
      - op: replace
        path: /build/artifacts/0/image
        value: "application-dev"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/image.repository
        value: "{{.IMAGE_FULLY_QUALIFIED_application_dev}}"
      - op: replace
        path: /deploy/helm/releases/0/name
        value: "{{.PREFIX}}-application"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/fullnameOverride
        value: "{{.PREFIX}}-application"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/secretName
        value: "{{.PREFIX}}-dbsecrets"
      - op: replace
        path: /deploy/helm/releases/0/setValueTemplates/configMapName
        value: "{{.PREFIX}}-config"
      - op: replace
        path: /build/local/push
        value: true
```

My dependency config file `dev-env.skaffold.yaml`:
```yaml
apiVersion: skaffold/v4beta7
kind: Config
metadata:
  name: dev-env
manifests:
  kustomize:
    paths:
      - redacted
deploy:
  kubectl:
    defaultNamespace: development
profiles:
  - name: dev

```

**Comment by renzodavid9** [comment]
Hey @gregwhorley, glad to hear that you were able to make it work. Oh! Yes, you have a profile in your config. Just for future references, something that I think may have worked too is to have the dev profile in your `dev-env.skaffold.yaml` file, so you can add the `requires` at the top level of your main config (not inside a profile), something like this:

main `skaffold.yaml`:
```
apiVersion: skaffold/v4beta7
kind: Config

requires:
  - path: ./dev-env.skaffold.yaml

profiles:
  - name: dev
    deploy:
      helm:
        releases: ...
```
And your `dev-env.skaffold.yaml`:
```
apiVersion: skaffold/v4beta7
kind: Config
metadata:
  name: dev-env

manifests: {}

profiles:
  - name: dev
    manifests:
      kustomize:
        paths:
          - redacted
    deploy:
      kubectl:
        defaultNamespace: development
```
Running `skaffold dev -p dev` in your main `skaffold.yaml` will activate the profile in both configs. I'm going to close this issue, please feel free to re open or open a new one if needed. Thanks!
